cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/DataprocUtils.java

```diff
@@ -350,19 +350,41 @@ public static void emitMetric(ProvisionerContext context, DataprocMetric datapro
   private static String getMetadata(String resource) throws IOException {
     URL url = new URL("http://metadata.google.internal/computeMetadata/v1/" + resource);
     HttpURLConnection connection = null;
-    try {
-      connection = (HttpURLConnection) url.openConnection();
-      connection.setRequestProperty("Metadata-Flavor", "Google");
-      connection.connect();
-      try (Reader reader = new InputStreamReader(connection.getInputStream(),
-          StandardCharsets.UTF_8)) {
-        return CharStreams.toString(reader);
+    ExponentialBackOff backOff = new ExponentialBackOff.Builder()
+      .setInitialIntervalMillis(MIN_WAIT_TIME_MILLISECOND)
+      .setMaxIntervalMillis(MAX_WAIT_TIME_MILLISECOND).build();
+
+    int counter = 0;
+    Exception exception = null;
+    while (counter < NUMBER_OF_RETRIES) {
+      counter++;
+
+      try {
+        connection = (HttpURLConnection) url.openConnection();
+        connection.setRequestProperty("Metadata-Flavor", "Google");
+        connection.connect();
+        try (Reader reader = new InputStreamReader(connection.getInputStream(),
+                                                   StandardCharsets.UTF_8)) {
+          return CharStreams.toString(reader);
+        }
+      } catch (Exception e) {
```

Comment:
we do not need to retry for every exception, we should only retry 5xx.

Reply:
Acknowledged